### PR TITLE
Added variantViz.py and some tests

### PR DESCRIPTION
**Pull request by jpdna**
Variant based visualizations for Mango and some related test.

Todo: 
1)  Additional tests are needed to cover all the variant visualization functions
2)  Create a smaller test data file will be helpful to speed testing time, replacing exome_chr22.adam with a smaller region or sampling from it.
3) Would like to add back in eQTL visualizations in the future, but possibly as a different future PR.

### CODE REVIEW

File: mango-python/bdgenomics/mango/variantViz.py [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+import matplotlib.ticker as ticker
+import math
+from pyspark.sql import Row
+plt.rcdefaults()
+plt.rcParams['figure.figsize'] = [10, 8]
+
+"""
+so_terms in order of least to most clinically significant
+"""
+so_terms = [
```
Comment (akmorrow13) [inline mango-python/bdgenomics/mango/variantViz.py]: Should this list go in ADAM or somewhere else?

File: mango-python/bdgenomics/mango/variantViz.py [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+            lambda x: x['attributes'][pop2] != ".").map(lambda d: ((float(d['attributes'][pop1])), float(d['attributes'][pop2]))   ).collect()
+        self.x = [ i[0] for i in curr ]
+        self.y = [ i[1] for i in curr ]
+        self.pop1 = pop1
+        self.pop2 = pop2
+
+    def plot(self):
+        plt.rcdefaults()
+        plt.title("Population Frequency\ncomparing populations " + self.pop1 + " and " + self.pop2)
+        mag = [ math.log(max( ( (z[0]-z[1]/(z[0]+.000000000001)), (z[1]-z[0])/(z[1]+.000000000001)))+1) for z in zip(self.x,self.y) ]
```
Comment (akmorrow13) [inline mango-python/bdgenomics/mango/variantViz.py]: what is .000000000001?

File: mango-python/bdgenomics/mango/variantViz.py [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+
+class VariantEffectCounts(object):
+    """
+    Plot Barchart of Variant Effect Type Counts
+    """
+    def __init__(self, ss, variantRDD, includeList = []):
+        if not includeList:
+            includeList = [
+                u'non_coding_transcript_exon_variant',
+                u'3_prime_UTR_variant',
```
Comment (akmorrow13) [inline mango-python/bdgenomics/mango/variantViz.py]: This seems to be redundant to the list above

File: mango-python/bdgenomics/mango/variantViz.py [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+          plt.xlabel("Number of Variants")
+          plt.ylabel("Number of Genes")
+          plt.show()
+        return self.data
+
+class VariantsGenomicRegion(object):
+    """
+    Plot variant density across chromosome region
+    """
+    def __init__(self,ss,variantRDD, start = 0, end = 1000000000, contigName="22"):
```
Comment (akmorrow13) [inline mango-python/bdgenomics/mango/variantViz.py]: why are these set? start = 0, end = 1000000000, contigName="22"

File: mango-python/bdgenomics/mango/variantViz.py [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+        self.annot_list = annot_list
+        self.bins = bins
+        self.sc = ss.sparkContext
+
+    def plot(self, testMode = False):
+        if (not testMode):
+          plt.rcdefaults()
+          plt.title("Allele Frequency Distribution by Functional Category")
+          bins = [.00001,.0001,.01,.025,.05,.075,.1,.15,.2,.25,.3,.35,.4,.45,.5]
+          #plt.hist(self.data_bins, weights=self.data_weights, bins=500, normed=1, histtype = 'step', cumulative=1, label=self.annot_list)
```
Comment (akmorrow13) [inline mango-python/bdgenomics/mango/variantViz.py]: remove comments

File: mango-python/bdgenomics/mango/variantViz.py [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+    Plot Allele Frequency Distribution by Population
+    """
+    def __init__(self, ss, variantRDD, pop_list=[ 'AF_NFE', 'AF_AFR']):
+        """
+        :param ss: SparkContext 
+        :param variantRDD: bdgenomics VariantRDD
+        :param pop_list: list of populations
+        """
+
+        def jp3(dd1):
```
Comment (akmorrow13) [inline mango-python/bdgenomics/mango/variantViz.py]: function name?

File: mango-python/bdgenomics/mango/variantViz.py [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+            effectList = [ effect for transcript in dd1['transcriptEffects'] for effect in transcript['effects'] ]
+            mostSigEffect = most_sig_effect(effectList)
+            return mostSigEffect
+
+        data_bins = []
+        data_weights = []
+
+        def getmaf(x):
```
Comment (akmorrow13) [inline mango-python/bdgenomics/mango/variantViz.py]: add function definition

File: mango-python/bdgenomics/mango/variantViz.py [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+    """
+    def __init__(self,ss,variantRDD, start = 0, end = 1000000000, contigName="22"):
+        """
+        :param ss: SparkContext 
+        :param variantRDD: bdgenomics variantRDD
+        :param start: genomic start position
+        :param end:  genomic stop position
+        :param contigName: contig name
+        """
+        data = variantRDD.toDF().filter((variantRDD.toDF().start > start) & (variantRDD.toDF().start < end) & (variantRDD.toDF().contigName == contigName) ).rdd.map(lambda w: w.asDict()).map(lambda x: (x['contigName'], int(math.floor(float(x['start'])/1000000)))).countByValue()
```
Comment (akmorrow13) [inline mango-python/bdgenomics/mango/variantViz.py]: divy up line

File: mango-python/bdgenomics/mango/variantViz.py [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+    """
+    Plot Histogram of variant count per sample
+    """
+    def __init__(self,ss,genotypeRDD):
+        """
+        :param ss: SparkContext
+        :param genotypeRDD: bdgenomics genotypeRDD 
+        """
+        df =  genotypeRDD.toDF()
+        data =  df.where( (df.alleles[0] == u'ALT') | (df.alleles[1] == u'ALT')).groupBy("sampleid").count().collect()
```
Comment (akmorrow13) [inline mango-python/bdgenomics/mango/variantViz.py]: where is "sampleid" defined?

File: mango-python/bdgenomics/mango/variantViz.py [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+        """
+        :param ss: SparkContext 
+        :param genotypeRDD: bdgenomics genotypeRDD
+        """
+        df =  genotypeRDD.toDF()
+        data =  df.where( ( ( len(df.alternateAllele ) > 1 )) & (  (df.alleles[0] == u'ALT') | (df.alleles[1] == u'ALT'))).groupBy("sampleid").count().collect()
+        counts = [ sample.asDict()['count'] for sample in data ]
+        self.counts = counts
+    def plot(self):
+        plt.hist(self.counts,bins=100)
```
Comment (akmorrow13) [inline mango-python/bdgenomics/mango/variantViz.py]: bins=100 is used a lot. Define it as a variable somewhere

File: mango-python/bdgenomics/mango/variantViz.py [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+class QDDist(object):
+    """
+    Plot Histogram of Quality By Depth (QD) score from a variant dataset
+    """
+    def __init__(self,ss,variantRDD):
+        """
+        :param ss: SparkContext 
+        :param variantRDD: bdgenomics variantRDD
+        """
+        data = variantRDD.toDF().rdd.map(lambda w: w.asDict()).map(lambda p: p['annotation']).map(lambda x: math.floor(float(x['attributes']['QD'])*10/10)).countByValue()
```
Comment (akmorrow13) [inline mango-python/bdgenomics/mango/variantViz.py]: why *10/10?
make new line

File: mango-python/bdgenomics/mango/variantViz.py [hunk 12]
```diff
@@ -0,0 +1,442 @@
...
+        self.scores = [ x[0] for x in data.items() ]
+        self.weights = [ x[1] for x in data.items() ]
+
+    def plot(self, testMode = False):
+        if (not testMode):
+          plt.hist(self.scores,weights=self.weights, bins=80)
```
Comment (akmorrow13) [inline mango-python/bdgenomics/mango/variantViz.py]: why bins=80?

File: mango-python/bdgenomics/mango/variantViz.py [hunk 13]
```diff
@@ -0,0 +1,442 @@
...
+
+        self.data_bins = data_bins
+        self.data_weights = data_weights
+        self.pop_list=pop_list
+        self.sc = ss.sparkContext
+
+    def plot(self):
+        plt.rcdefaults()
+        plt.title("Allele Frequency Distribution by Population")
+        plt.hist(self.data_bins, weights=self.data_weights, bins=[.00001,.0001,.001,.01,.025,.05,.075,.1,.15,.2,.25,.3,.35,.4,.45,.5], histtype = 'step', label=self.pop_list)
```
Comment (akmorrow13) [inline mango-python/bdgenomics/mango/variantViz.py]: what is the bin list for?

File: mango-python/bdgenomics/mango/variantViz.py [hunk 14]
```diff
@@ -0,0 +1,442 @@
...
+        def getmaf(x):
+            if(x<0.5):
+                return x
+            else:
+              return 1 - x
+
+        for pop in pop_list:
+           curr = variantRDD.toDF().rdd.map(lambda w: w.asDict()).map(lambda p: p['annotation']).filter(
+             lambda x: x['attributes'][pop] != ".").map(
+             lambda d: (jp3(d), math.floor( getmaf(float(d['attributes'][pop])) *100000)/100000)).filter(lambda f: f[1] > 0).countByValue()
```
Comment (akmorrow13) [inline mango-python/bdgenomics/mango/variantViz.py]: remove *100000)/100000

File: mango-python/bdgenomics/mango/variantViz.py [hunk 15]
```diff
@@ -0,0 +1,442 @@
...
+import matplotlib.pyplot as plt
+import numpy as np
+import matplotlib.ticker as ticker
+import math
+from pyspark.sql import Row
+plt.rcdefaults()
+plt.rcParams['figure.figsize'] = [10, 8]
+
+"""
+so_terms in order of least to most clinically significant
```
Comment (akmorrow13) [inline mango-python/bdgenomics/mango/variantViz.py]: link to significance

File: mango-python/bdgenomics/mango/variantViz.py (hunk 4 above)
Comment (jpdna) [reply mango-python/bdgenomics/mango/variantViz.py]: I mean it as an example of approrpiate parameterws, but you are right, it really makes more sense to have no default, I will get rid of

File: mango-python/bdgenomics/mango/variantViz.py [hunk 16]
```diff
@@ -0,0 +1,442 @@
...
+        if len(annot_list) == 0:
+          annot_list = ['missense_variant', 'synonymous_variant']
+
+        data_bins = []
+        data_weights = []
+
+        for annot in annot_list:
+              curr = variantRDD.toDF().rdd.map(lambda w: w.asDict()).map(lambda p: p['annotation']).filter(
+                  lambda x: x['attributes']['AF_NFE'] != ".").map(
+                  lambda d: (get_effect(d), math.floor(float(d['attributes']['AF_NFE'])*10000)/10000)).filter(
```
[the remark this answers is not in the document]
Comment (jpdna) [reply mango-python/bdgenomics/mango/variantViz.py]: You are right, I made it a parameter

File: mango-python/bdgenomics/mango/variantViz.py (hunk 3 above)
Comment (jpdna) [reply mango-python/bdgenomics/mango/variantViz.py]: Its not the same list, this is a default list that will give a meaningful view of many gene related variant effect counts without without being dominated by intergenic and intronic types.  The user can overide this parameter to get other categories if they choose.

File: mango-python/bdgenomics/mango/variantViz.py (hunk 7 above)
Comment (jpdna) [reply mango-python/bdgenomics/mango/variantViz.py]: done

File: mango-python/bdgenomics/mango/variantViz.py (hunk 8 above)
Comment (jpdna) [reply mango-python/bdgenomics/mango/variantViz.py]: done

File: mango-python/bdgenomics/mango/variantViz.py (hunk 12 above)
Comment (jpdna) [reply mango-python/bdgenomics/mango/variantViz.py]: It seemed to make a readable histogram for a typical distribution of this QD value

File: mango-python/bdgenomics/mango/variantViz.py (hunk 9 above)
Comment (akmorrow13) [reply mango-python/bdgenomics/mango/variantViz.py]: It is a little strange to hard code it as a string if it is part of a schema

File: mango-python/bdgenomics/mango/variantViz.py (hunk 12 above)
Comment (akmorrow13) [reply mango-python/bdgenomics/mango/variantViz.py]: should it be an option?

File: mango-python/bdgenomics/mango/variantViz.py (hunk 9 above)
Comment (jpdna) [reply mango-python/bdgenomics/mango/variantViz.py]: Is there an alternative way to construct groupBy caluse  this without hardcoding it?

**Comment by shaneknapp** [comment]
test this please (for ubuntu port)

**Comment by shaneknapp** [comment]
test this please

**Comment by jpdna** [comment]
closed in favor of #427

**Pull request closed without merging**